### PR TITLE
Ethers v6 + various refactorings

### PR DESCRIPTION
**Pull request by CedarMist**
In summary, as there are quite a few changes:

 * upgrade to Ethers v6
 * split typechain generated types into an ESM package (`contracts`) and the underlying contracts + hardhat build stuff into the `hardhat` package
 * removed `functions`, there's no API or workers now
 * tested with sapphire-paratime master (1.3.x)
 * added Makefiles
 * made UI more robust
 * updated dependencies

### CODE REVIEW

**Comment by lubej** [comment]
Lets wait for the https://github.com/oasisprotocol/sapphire-paratime/pull/261 to be merged, so that the dependency can be updated as well. And afterwards I will look into the PR changes, I missed the previous PR refactoring merge.

**Comment by CedarMist** [comment]
> Lets wait for the [oasisprotocol/sapphire-paratime#261](https://github.com/oasisprotocol/sapphire-paratime/pull/261) to be merged, so that the dependency can be updated as well. And afterwards I will look into the PR changes, I missed the previous PR refactoring merge.

I have published the two packages and updated the dependencies.

Now with the peer dependencies the resulting build is fairly small

image

File: pnpm-workspace.yaml [hunk 1]
```diff
@@ -1 +1 @@
-packages: [frontend, backend, functions]
+packages: [hardhat, contracts,  functions, frontend]
```
Comment (aefhm) [inline pnpm-workspace.yaml]: (nit)
```suggestion
packages: [hardhat, contracts, functions, frontend]
```

File: hardhat/src/index.ts [hunk 2]
```diff
@@ -1,8 +1,10 @@
...
 export {
     DAOv1, DAOv1__factory,
     PollACLv1, PollACLv1__factory,
     WhitelistVotersACLv1, WhitelistVotersACLv1__factory,
     AllowAllACLv1, AllowAllACLv1__factory,
     AllowVotersACLv1, AllowVotersACLv1__factory,
     GaslessVoting, GaslessVoting__factory
-} from "../typechain-types"
+} from "./contracts"
+*/
```
Comment (aefhm) [inline hardhat/src/index.ts]: (nit)
```suggestion
*/

```

File: hardhat/contracts/AcceptsProxyVotes.sol [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { ProposalId, ProposalParams } from "./Types.sol";
+import { PollACLv1 } from "./PollACLv1.sol";
+
+interface AcceptsProxyVotes {
```
Comment (aefhm) [inline hardhat/contracts/AcceptsProxyVotes.sol:7]: Maybe we can remove this later too. (not requested in this PR).

File: frontend/vite.config.mts [hunk 4]
```diff
@@ -3,9 +3,11 @@ import { fileURLToPath, URL } from 'node:url';
 import { defineConfig } from 'vite';
 import vue from '@vitejs/plugin-vue';
 import { visualizer } from 'rollup-plugin-visualizer';
+//import { viteSingleFile } from "vite-plugin-singlefile"
```
Comment (aefhm) [inline frontend/vite.config.mts]: Remove?

File: frontend/src/views/HomeView.vue [hunk 5]
```diff
@@ -1,11 +1,11 @@
 <script setup lang="ts">
-import { ethers } from 'ethers';
+import { ethers } from "ethers";
```
Comment (aefhm) [inline frontend/src/views/HomeView.vue]:
```suggestion
import { ethers } from 'ethers';
```

File: frontend/src/views/PollView.vue [hunk 6]
```diff
@@ -197,16 +203,16 @@ onMounted(async () => {
             v-for="(choice, choiceId) in poll.ipfsParams.choices"
             :key="choiceId"
             :class="{
-              selected: selectedChoice === choiceId || choiceId === winningChoice,
+              selected: Number(selectedChoice) === choiceId || choiceId === Number(winningChoice),
```
Comment (aefhm) [inline frontend/src/views/PollView.vue:206]: Number?

**Review by aefhm — APPROVED** [review]

File: frontend/src/views/PollView.vue (hunk 6 above)
Comment (CedarMist) [reply frontend/src/views/PollView.vue:206]: Yea, the contract's returning bigints the JS file contains Numbers, one needs to be cast to another for comparison. There is probably a cleaner way.

File: frontend/vite.config.mts (hunk 4 above)
Comment (CedarMist) [reply frontend/vite.config.mts]: I was looking at reducing loading times on slow connections

The vite-plugin-singlefile is nice, but you can't control the ordering.

Ideally I'd have the CSS, then the page HTML, then the ~500kb blob of Javascript (200kb compressed) all in one file, that way the client loads the CSS and we can have a nice styled 'loading page' in the default HTML, then when the rest of the app loads the app appears. I've found this works really well in practice as long as you can keep the JS file size down to ~200-300kb (compressed).

I'll see if I can get it down to just 2 files, so CSS is inlined in the HTML file, then the JS is loaded separately.

**Pull request merged**